### PR TITLE
[Sol->Yul] Evaluate base arguments in derived to base order.

### PR DESCRIPTION
**Pull request by chriseth**

### CODE REVIEW

**Comment by mijovic** [comment]
It looks good. 
Do you think it makes sense to add more tests for this, with some side effects, so we can make sure this is correct?

**Comment by ekpyron** [comment]
Is this the best solution? IIUC this will still mean that modifier and base constructor arguments in a list like
```
constructor() mod(f()) Base1(g()) Base3(h()) Base2(i()) mod2(j()) {}
```
will be allowed and have a quite hard to understand order, doesn't it?

The reason for not just evaluating left-to-right was that base constructor arguments need to be evaluated first and modifier arguments only later or something like that, right? (Not sure anymore and not thinking it through entirely again now).

If that's the case wouldn't it make more sense to just require inheritance specifiers to come before modifiers syntactically (this holds for old and for new codegen anyways, right?) and then just evaluate left-to-right *always*?

To me that would be a much clearer rule... with the current proposal, even if I know the rules, I can't immediately see the evaluation order until I look around globally...

All that being said, I'd probably also be fine with this way of doing things :-).

**Comment by ekpyron** [comment]
So I'd argue restricting the order of inheritance specifiers vs. modifiers for constructors may be a valuable breaking change to yet introduce in 0.8... but I'm also not entirely sure about it...

**Comment by chriseth** [comment]
This PR just gets the Sol->Yul implementations closer to our current Sol->EVM implementation. Whether or not that is a good thing should be evaluated as a breaking change for 0.9.x

**Comment by ekpyron** [comment]
> This PR just gets the Sol->Yul implementations closer to our current Sol->EVM implementation. Whether or not that is a good thing should be evaluated as a breaking change for 0.9.x

Ok, that's fair.

File: libsolidity/codegen/ir/IRGenerator.cpp [hunk 1]
```diff
@@ -437,17 +437,32 @@ pair<string, map<ContractDefinition const*, vector<string>>> IRGenerator::evalua
...
+	struct InheritanceOrder
+	{
+		bool operator()(ContractDefinition const* _c1, ContractDefinition const* _c2) const
+		{
+			solAssert(contains(linearizedBaseContracts, _c1) && contains(linearizedBaseContracts, _c2), "");
+			auto it1 = find(linearizedBaseContracts.begin(), linearizedBaseContracts.end(), _c1);
+			auto it2 = find(linearizedBaseContracts.begin(), linearizedBaseContracts.end(), _c2);
+			return it1 < it2;
+		}
+		std::vector<ContractDefinition const*>& linearizedBaseContracts;
```
Comment (ekpyron) [inline libsolidity/codegen/ir/IRGenerator.cpp]: Why is this a non-const reference?

Comment (ekpyron) [reply libsolidity/codegen/ir/IRGenerator.cpp]: Also could loose the ``std::``.

Comment (chriseth) [reply libsolidity/codegen/ir/IRGenerator.cpp]: Fixed.

**Review by ekpyron — DISMISSED** [review]
It would be good to have test cases for this, if we don't already (apparently at least none pop up as now being available with yul...)... but fine to merge without adding them here...

**Comment by chriseth** [comment]
Added test.

**Pull request merged**